src/Core/Module/ModuleRepository.php

```diff
@@ -0,0 +1,249 @@
+    private $adminModuleDataProvider;
+
+    /** @var HookManager */
+    private $hookManager;
+
+    /** @var CacheProvider */
+    private $cacheProvider;
+
+    /** @var string */
+    private $modulePath;
+
+    public function __construct(
+        ModuleDataProvider $moduleDataProvider,
+        AdminModuleDataProvider $adminModuleDataProvider,
+        CacheProvider $cacheProvider,
+        HookManager $hookManager,
+        string $modulePath
+    ) {
+        $this->finder = new Finder();
+        $this->moduleDataProvider = $moduleDataProvider;
+        $this->adminModuleDataProvider = $adminModuleDataProvider;
+        $this->cacheProvider = $cacheProvider;
+        $this->hookManager = $hookManager;
+        $this->modulePath = $modulePath;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            ModuleManagementEvent::INSTALL => 'onModuleStateChanged',
+            ModuleManagementEvent::UNINSTALL => 'onModuleStateChanged',
+            ModuleManagementEvent::ENABLE => 'onModuleStateChanged',
+            ModuleManagementEvent::DISABLE => 'onModuleStateChanged',
+            ModuleManagementEvent::UPGRADE => 'onModuleStateChanged',
+            ModuleManagementEvent::ENABLE_MOBILE => 'onModuleStateChanged',
+            ModuleManagementEvent::DISABLE_MOBILE => 'onModuleStateChanged',
+        ];
+    }
+
+    public function onModuleStateChanged(ModuleManagementEvent $event)
+    {
+        $moduleName = $event->getModule()->get('name');
+        if ($this->cacheProvider->contains($moduleName)) {
+            $this->cacheProvider->delete($moduleName);
+        }
+    }
+
+    public function getList(): array
+    {
+        $modules = [];
+        $modulesDirsList = $this->finder->directories()
+            ->in($this->modulePath)
+            ->depth('== 0')
+            ->exclude(['__MACOSX'])
+            ->ignoreVCS(true);
+
+        foreach ($modulesDirsList as $moduleDir) {
+            $moduleName = $moduleDir->getFilename();
+            if (!file_exists($this->modulePath . $moduleName . '/' . $moduleName . '.php')) {
+                continue;
+            }
+
+            $modules[] = $this->getModule($moduleName);
+        }
+
+        return $this->mergeWithModulesFromHook($modules);
+    }
+
+    public function getInstalledModules(): array
+    {
+        return array_filter($this->getList(), function ($module) {
+            return $module->database->get('installed') === 1;
+        });
+    }
+
+    public function getConfigurableModules(): array
+    {
+        return array_filter($this->getList(), function (Module $module) {
+            return $module->hasValidInstance() && !empty($module->getInstance()->warning);
+        });
+    }
+
+    public function getUpgradableModules(): array
+    {
+        return array_filter($this->getList(), function (Module $module) {
+            return $module->canBeUpgraded();
+        });
+    }
+
+    /**
+     * @return Module
```

Comment:
PhpDoc is useless here, especially when inconsistent with PHP return type :)
```suggestion
     * @return ModuleInterface
```